src/apps/projinfo.cpp

```diff
@@ -1248,6 +1263,76 @@ int main(int argc, char **argv) {
         }
     }
 
+    if (!listCRS.empty()) {
+        auto params = proj_get_crs_list_parameters_create();
+        bool all_types = false;
+        std::vector<PJ_TYPE> types;
+        std::vector<std::string> auth_names;
+        auto tokens = split(listCRS, ',');
+        for (auto token : tokens) {
+            auto auth_tokens = split(token, '=');
+            if (auth_tokens.size() == 2 && ci_equal(auth_tokens[0], "auth_name")) {
+                auth_names.push_back(auth_tokens[1]);
+                continue;
+            }
+
+            if (ci_equal(token, "allow_deprecated")) {
+                params->allow_deprecated = true;
+            } else if (ci_equal(token, "all_types")) {
+                all_types = true;
+            } else if (ci_equal(token, "geodetic")) {
+                types.push_back(PJ_TYPE_GEODETIC_CRS);
+            } else if (ci_equal(token, "geocentric")) {
+                types.push_back(PJ_TYPE_GEOCENTRIC_CRS);
+            } else if (ci_equal(token, "geographic")) {
+                types.push_back(PJ_TYPE_GEOGRAPHIC_CRS);
+            } else if (ci_equal(token, "geographic_2d")) {
+                types.push_back(PJ_TYPE_GEOGRAPHIC_2D_CRS);
+            } else if (ci_equal(token, "geographic_3d")) {
+                types.push_back(PJ_TYPE_GEOGRAPHIC_3D_CRS);
+            } else if (ci_equal(token, "vertical")) {
+                types.push_back(PJ_TYPE_VERTICAL_CRS);
+            } else if (ci_equal(token, "projected")) {
+                types.push_back(PJ_TYPE_PROJECTED_CRS);
+            } else if (ci_equal(token, "compound")) {
+                types.push_back(PJ_TYPE_COMPOUND_CRS);
+            } else if (ci_equal(token, "temporal")) {
+                types.push_back(PJ_TYPE_TEMPORAL_CRS);
+            } else if (ci_equal(token, "engineering")) {
+                types.push_back(PJ_TYPE_ENGINEERING_CRS);
+            } else if (ci_equal(token, "bound")) {
+                types.push_back(PJ_TYPE_BOUND_CRS);
+            } else if (ci_equal(token, "other")) {
+                types.push_back(PJ_TYPE_OTHER_CRS);
+            } else {
+                std::cerr << "Unrecognized value for option --list-crs: " << token
+                          << std::endl;
+                usage();
+            }
+        }
+
+        if (all_types) {
+            types.clear();
+        }
+        params->typesCount = types.size();
+        params->types = types.data();
+
+        if (auth_names.empty()) {
+            auth_names.push_back("");
+        }
+
+        for (auto auth_name : auth_names) {
+            int result_count = 0;
+            auto list = proj_get_crs_info_list_from_database(nullptr, auth_name.c_str(), params, &result_count);
```

Comment:
you should likely use proj_context_create() and proj_context_set_database_path() to take into account --main-db-path and --aux-db-path

Reply:
I understand that there is not way to extract the `PJ_CONTEXT*` from `DatabaseContextPtr dbContext` that is already created.

Reply:
> I understand that there is not way to extract the `PJ_CONTEXT*` from `DatabaseContextPtr dbContext` that is already created.

not reallly indeed. A DatabaseContext could potentially exist without being linked to a PJ_CONTEXT* (historically the C++ part of the codebase has been developed quite independently from the C part, hence they are loosely decoupled). An alternative might be that you look at the implementation of proj_get_crs_info_list_from_database() to still from it what you need. The C function largely uses the C++ API and thus you could re-use the dbContext already constructed

Reply:
Using the code from C++ implementation is a bit ugly: repeating the logic to filter the types, bbox, etc.
Maybe I try later to see how it looks like, in a different branch to compare.